### DISCUSSION TITLE
Does RabbitMQ server [partially] support downgrade?

### DISCUSSION THREAD

**Discussion by inikulshin** [rabbitmq/rabbitmq-server#12124] [Questions] [answered]
I understand that the generic answer for this generic question is 'no'.

But maybe there are only **few** well-known breaking versions, that can't be skipped while downgrading, and besides those versions downgrade is allowed?

E.g. I suppose that downgrades from `3.X.Y` to `3.X.Y-1` will work for almost every X and Y. But what about downgrade from 3.X to 3.X-1?

* We are interested in this knowledge to improve upgrade/rollback implementation - for versions that can be downgraded, no node backup is required before upgrade.

**Comment by michaelklishin** [comment, marked as answer]
We have previously explained in several different discussions that some data migrations are irreversible inherently, and others are irreversible because our small team won't spend our time on them.

They are rare for patch releases but can still happen. They are fairly common between feature/minor releases. And for 4.0, it's a guarantee that a number of important internal changes was lined up as it is a new major release in 12 years.

[Blue-Green deployment](https://www.rabbitmq.com/docs/blue-green-upgrade) should be used for environments that have reasons to believe they may need to downgrade. In fact, Blue-Green deployments is the only safe way to downgrade as well, at least as long as the older cluster would accept a definition file from the newer one. Which is fairly likely.

**Comment by michaelklishin** [comment]
@inikulshin our team won't spend more time on any downgrading questions. You know the answer. You know about Blue-Green deployments.

Anyone who wants to test downgrades will have to do them on their own and face the consequences (primarily of in-place downgrades) without any support from the core team.
